server/etcdserver/bootstrap.go

```diff
@@ -64,61 +56,80 @@ func bootstrap(cfg config.ServerConfig) (b *bootstrappedServer, err error) {
 	if terr := fileutil.TouchDirAll(cfg.DataDir); terr != nil {
 		return nil, fmt.Errorf("cannot access data directory: %v", terr)
 	}
-
-	haveWAL := wal.Exist(cfg.WALDir())
 	ss := bootstrapSnapshot(cfg)
+	prt, err := rafthttp.NewRoundTripper(cfg.PeerTLSInfo, cfg.PeerDialTimeout())
+	if err != nil {
+		return nil, err
+	}
+
+	if terr := fileutil.TouchDirAll(cfg.MemberDir()); terr != nil {
+		return nil, fmt.Errorf("cannot access member directory: %v", terr)
+	}
 
-	be, ci, beExist, beHooks, err := bootstrapBackend(cfg)
+	st := v2store.New(StoreClusterPrefix, StoreKeysPrefix)
+	backend, err := serverstorage.BootstrapBackend(cfg, st, ss)
 	if err != nil {
 		return nil, err
 	}
-	defer func() {
-		if err != nil {
-			be.Close()
+	var (
+		bwal *serverstorage.BootstrappedWAL
+	)
+
+	if wal.Exist(cfg.WALDir()) {
```

Comment:
Can we move these logic to `bootstrapCluster` method and add `*serverstorage.BootstrappedWAL` to `bootstrappedCluster`? I think then we can simplify `Finalize` method as well (lower number of parameters)?

Reply:
One of the goals of this refactor was to split `cluster` versus `storage` bootstrap. Sure currently it's true that if we kept `WAL` under `cluster` code might be simpler, but this should not be our direction. 

Logically I don't think `WAL` should be part of `Cluster`, I think it belongs to storage layer. Especially with introduction of storage schema which will can influences how both bbolt and WAL writes entries.

There is definite egg and chicken problem during bootstrap. Depending on whether cluster bootstraps from clean state or not, either we create Cluster from WAL or WAL from cluster. Still I think we should work torwards removing this circular dependency instead of shuffling all the code into one place. if not we will not get much further than 500+ line functions like we had before https://github.com/etcd-io/etcd/blob/v3.5.0/server/etcdserver/server.go#L330-L691